### PR TITLE
Initial support for the upcoming version of the Kindle app 

### LINKED ISSUE #4300: Tighten top level client check in NVDAObjects.IAccessible.isPresentableFocusAncestor
**Reported by jteh on 2014-07-17 04:19**
Google Chrome exposes ROLE_SYSTEM_CLIENT for some of its accessibles which don't have their own HWND. This seems odd given the name CLIENT, but MSDN is much less specific:
 Microsoft Active Accessibility uses this role as a default if there is a question about the role of a UI element.

We can tighten this by using IAccessibleIdentity to check for OBJID_CLIENT, CHILDID_SELF.

This will prevent extraneous reporting of two panes when focusing Google Chrome.

### PR DESCRIPTION
**Pull request by michaelDCurran**
Specifically:
    \* Browse mode allowing for moving through the content by character, word, line etc with the arrow keys.
    \* Automatic page turns when reaching the start/end of a page with the arrow keys / Say All.
    \* Automatic reporting of page labels or numbers if available
    \* Reporting of current location within a book (including percentage) via NVDA's report location command (NVDA+numpadDelete).

This PR contains various small changes to treeInterceptors and textInfos,handling page turns in braille, and fixes for QT5. Fixes #4300.

### CODE REVIEW

File: source/NVDAObjects/IAccessible/__init__.py [hunk 1]
```diff
@@ -1300,7 +1300,7 @@ def event_selectionWithIn(self):
 
 	def _get_isPresentableFocusAncestor(self):
 		IARole = self.IAccessibleRole
-		if IARole == oleacc.ROLE_SYSTEM_CLIENT and self.windowStyle & winUser.WS_SYSMENU:
+		if IARole == oleacc.ROLE_SYSTEM_CLIENT and self.event_objectID==winUser.OBJID_CLIENT and self.event_childID==0 and self.windowStyle & winUser.WS_SYSMENU:
```
Comment (jcsteh) [inline source/NVDAObjects/IAccessible/__init__.py]: I think we should have a comment in this block explaining this, as it's a bit obscure. Something like this:

```
# This is the root object of a top level window.
# #4300: We check the object and child ids as well because there can be "clients" other than the root.
```

File: source/treeInterceptorHandler.py [hunk 2]
```diff
@@ -144,22 +144,27 @@ def makeTextInfo(self,position):
...
 	def copy(self):
-		return self.__class__(self.obj,None,_rangeObj=self.innerTextInfo._rangeObj)
+		innerCopy=self.innerTextInfo.copy()
+		return self.__class__(self.obj,innerCopy)
 
 	def _get__rangeObj(self):
 		return self.innerTextInfo._rangeObj
 
 	def _set__rangeObj(self,r):
 		self.innerTextInfo._rangeObj=r
```
Comment (jcsteh) [inline source/treeInterceptorHandler.py]: Do we still need this `_rangeObj` property now that the constructor and copy don't depend on `_rageObj`?

File: source/appModules/kindle.py [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+		try:
+			self.rootNVDAObject.appModule.inPageTurn=True
+			return self.rootNVDAObject.turnPage(previous=previous)
+		finally:
+			self.rootNVDAObject.appModule.inPageTurn=False
+
+	def isAlive(self):
+		if not winUser.isWindow(self.rootNVDAObject.windowHandle):
+			return False
+		return True
```
Comment (jcsteh) [inline source/appModules/kindle.py]: This can be simplified to just:

`return winUser.isWindow(self.rootNVDAObject.windowHandle)`

File: source/appModules/kindle.py [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+import time
+import appModuleHandler
+import speech
+import sayAllHandler
+import eventHandler
+import api
+from scriptHandler import willSayAllResume, isScriptWaiting
+import controlTypes
+import treeInterceptorHandler
+import IAccessibleHandler
```
Comment (jcsteh) [inline source/appModules/kindle.py]: Unused import.

File: source/appModules/kindle.py [hunk 5]
```diff
@@ -0,0 +1,139 @@
...
+from cursorManager import ReviewCursorManager
+from browseMode import BrowseModeDocumentTreeInterceptor
+import textInfos
+from textInfos import DocumentWithPageTurns
+from NVDAObjects.IAccessible import IAccessible, IA2TextTextInfo, getNVDAObjectFromEvent
```
Comment (jcsteh) [inline source/appModules/kindle.py]: getNVDAObjectFromEvent is an unused import.

File: source/treeInterceptorHandler.py [hunk 6]
```diff
@@ -144,22 +144,27 @@ def makeTextInfo(self,position):
...
 class RootProxyTextInfo(textInfos.TextInfo):
 
-	def __init__(self,obj,position,_rangeObj=None):
+	def __init__(self,obj,position,**kwargs):
 		super(RootProxyTextInfo,self).__init__(obj,position)
-		self.innerTextInfo=self.InnerTextInfoClass(obj.rootNVDAObject,position,_rangeObj=_rangeObj)
-
-	def _get_InnerTextInfoClass(self):
-		return self.obj.rootNVDAObject.TextInfo
+		self.InnerTextInfoClass=self.obj.rootNVDAObject.TextInfo
```
Comment (jcsteh) [inline source/treeInterceptorHandler.py]: I'm concerned this will break Word, which seems to set InnerTextInfoClass to WordDocumentTextInfoForTreeInterceptor, which overrides shouldIncludeLayoutTables.

File: source/appModules/kindle.py [hunk 7]
```diff
@@ -0,0 +1,139 @@
...
+
+class BookPageViewTreeInterceptor(DocumentWithPageTurns,ReviewCursorManager,BrowseModeDocumentTreeInterceptor):
+
+	TextInfo=treeInterceptorHandler.RootProxyTextInfo
+
+	def turnPage(self,previous=False):
+		try:
+			self.rootNVDAObject.appModule.inPageTurn=True
```
Comment (jcsteh) [inline source/appModules/kindle.py]: Why is this needed? I think we need a comment here, possibly referring to the focus ignorer thingy.

File: source/appModules/kindle.py [hunk 8]
```diff
@@ -0,0 +1,139 @@
...
+			curTime=time.time()
+		else:
+			raise RuntimeError("no more pages")
+
+class PageTurnFocusIgnorer(IAccessible):
+
+	def _get_shouldAllowIAccessibleFocusEvent(self):
+		if self.appModule.inPageTurn:
+			return False
+		return super(PageTurnFocusIgnorer,self).shouldAllowIAccessibleFocusEvent
```
Comment (jcsteh) [inline source/appModules/kindle.py]: Why is this necessary? What extraneous focus gets fired during a page turn? Please comment.

File: source/treeInterceptorHandler.py (hunk 2 above)
Comment (michaelDCurran) [inline source/treeInterceptorHandler.py]: I think so, yes. There may be still code in UIA or MS word browse modes that may depend on accessing it. We should look into removing them if logically possible, but not in this particular PR. The change here was to specifically allow copying with no _rangeObj.

File: source/treeInterceptorHandler.py (hunk 6 above)
Comment (michaelDCurran) [inline source/treeInterceptorHandler.py]: I agree. I'll change that back.

**Comment by vgjh2005** [comment]
Hi Mic:
Can you provide the download link of this app? Thanks!

**Comment by jcsteh** [comment]
It's not released yet. No estimated release date has been given.
